Add next-gen beta documentation (#1565)
slackapi/bolt-js: docs/_future/create_new_app.md

To create the app, you'll run the following command:
```
slack create my-app -t slack-samples/bolt-js-request-time-off
```
This command creates an app through the CLI by cloning a specified template. In this case, the template is the [Bolt for JavaScript Request Time Off](https://github.com/slack-samples/bolt-js-request-time-off) application. 

The application will take a few moments to set up and clone the repository. Once the app is successfully created, you should see a message like this:
```
✨ my-app successfully created

🧭 Explore your project's README.md for documentation and code samples, and at any time run hermes help to display a list of available commands

🧑‍🚀 Follow the steps below to try out your new project

1️⃣  Change into your project directory with: cd my-app

2️⃣  Develop locally and see changes in real-time with: slack run

3️⃣  When you're ready to deploy for production use: slack deploy

🔔 If you leave the workspace, you won’t be able to manage any apps you’ve deployed to it. Apps you deploy will belong to the workspace even if you leave the workspace
```

You can now explore the `my-app` directory!

---
### Set up your trigger {#setup-trigger}

> In order to utilize the pre-existing functionality in the Request Time Off app, you'll need to create a trigger prior to running your app.

Reviewer: ```suggestion
This app comes setup with pre-existing functionality - Functions, Workflows and a Trigger set up to kick things off. Let's run a command to initialize that Trigger via the CLI
```

First, make sure you're in the project directory in your command line: `cd my-app`

> Then, run the following command to create a trigger using the `triggers/link-shortcut.json` configuration file. This file was brought into your project from the Bolt for JavaScript Request Time Off template and is used to initialize your trigger:

Reviewer: Suggestion: Let's switch the order around, have them run the command and see it working first, then explain what just happened and where the config file comes from and what the file is. Then we can also add a link to the triggers section of the docs at the end.

```
slack triggers create --trigger-def "triggers/link-shortcut.json"      
```

The above command will create a trigger for the selected workspace. Once the trigger is successfully created, you should see an output like this:
```
⚡ Trigger created
   Trigger ID:   [ID]
   Trigger Type: shortcut
   Trigger Name: Take Your Time

>    Shortcut URL: https://slack.com/shortcuts/[ID]/[Some ID]

Reviewer: ```suggestion
   URL: https://slack.com/shortcuts/[ID]/[Some ID]
```
Output is now just URL I believe. Inevitably we'll get drift on these, but that's the nature of docs I suppose

```
The provided URL will be what you use to run your trigger. Copy this URL and save it somewhere; you'll need it for later.

---

> ### Run your app {#run-your-app}

Reviewer: I'd break this into two sections and put everything starting L94 into the second header

### Run your app
### Trigger your app's workflow

With your app running, access your workspace and paste the URL from the the trigger you created in the previous step into a message in a public channel.

> 💡 To make the trigger URL more widely accessible, we recommend saving the trigger as a channel bookmark for easy access.

Send the message and click the "Run" button that appears. A modal will appear prompting you to enter information to request time off. To test your app properly, we recommend entering your own Slack username in the "Manager" field.

![Request Time Off modal](../assets/take-your-time-modal.png "Request Time Off modal")

Then, submit the form. You should receive a message from the app with information about the requested time off as well as an "Approve" and "Deny" button. Once either button is selected, you should then receive a message from the app notifying you of the status of your request.

The full app flow can be seen here:
![Request Time Off app](../assets/take-your-time-demo.gif "Request Time Off app")

> Now that we have a working instance of the app in your workspace, let's dive into its components and how it works!

Reviewer: Let's give the option to jump ship here: 

```suggestion
Now we have a working instance of the app in your workspace and you've seen it in action! You can exploring on your own and dive into the code yourself here or follow along as we break down this app's next-gen components below to see how everything works!
```

---
### Workflows {#workflows}

Workflows are a set of processing steps that are executed in order. Every workflow step consists of either a [built-in function](/bolt-js/future/built-in-functions) or [custom function](/bolt-js/future/custom-functions). Workflows can be configured to run without any user input, or they can wait for input via form before continuing.

In the Request Time Off application, the Time Off Request workflow is defined in `manifest/workflows/time-off-request.js`. The general structure of a workflow is to first define it using the built in function `DefineWorkflow`:

```js
const { DefineWorkflow, Schema } = require('@slack/bolt');

const TimeOffWorkflow = DefineWorkflow({
  callback_id: 'time_off_request_wf',
  title: 'Time Off Request Workflow',
  input_parameters: {
    properties: {
      interactivity: {
        type: Schema.slack.types.interactivity,
      },
    },
    required: [],
  },
});

module.exports = { TimeOffWorkflow };
```

> In order to create a workflow that utilizes user input, such as one that uses a form, you'll need to pass in `interactivity` as one of the workflow's `input_parameters` under `properties`. This will allow the form to be interacted with.

Reviewer: ```suggestion
Here we want our workflow to take in user input, like form data. To do this we will need to pass in `interactivity` as one of the workflow's `input_parameters` under `properties`. This will allow the form to be interacted with.
```

> Now that the workflow has been defined, you can start adding steps underneath the definition, which will execute additional functionality in your workflow. Steps are executed in the order they appear. For the Time Off Request workflow, the first step will be a form that will allow the user to enter in their time off request.

Reviewer: ```suggestion
With our workflow defined, we can start by adding steps to it, which will execute additional functionality in your workflow. Steps are executed in the order you defined them. For the Time Off Request workflow, the first step will be a form that will allow the user to enter in their time off request.
```

```js
const step1 = TimeOffWorkflow.addStep(
  Schema.slack.functions.OpenForm,
  {
    title: 'Request Time Off',
    submit_label: 'Request',
    description: 'Please describe your request',
    interactivity: TimeOffWorkflow.inputs.interactivity,
    fields: {
      required: ['manager', 'start_date', 'end_date'],
      elements: [
        {
          name: 'manager',
          title: 'Manager',
          type: Schema.slack.types.user_id,
        },
        {
          name: 'start_date',
          title: 'Start Date',
          type: Schema.slack.types.timestamp,
        },
        {
          name: 'end_date',
          title: 'End Date',
          type: Schema.slack.types.timestamp,
        },
      ],
    },
  },
);
```

> This step uses the built-in function [OpenForm](https://api.slack.com/future/functions#open-a-form) and initializes expected inputs of the form, which are defined by [Slack types](https://api.slack.com/future/types), such as the `timestamp` type, which represents dates and times. The Time Off Request form contains a `manager`, `start_date`, and `end_date` field, all of which are required inputs to submit the form.

Reviewer: ```suggestion
In the past, to open a form in a modal, you'd use our `client.views.open` method and pass it some custom Block Kit for formatting. Now you can simplify this step with a built-in function, OpenForm and Slack will take care of the formatting. 

This step uses a Slack built-in function [OpenForm](https://api.slack.com/future/functions#open-a-form) and initializes expected inputs of the form, which are defined by [Slack types](https://api.slack.com/future/types), such as the `timestamp` type, which represents dates and times. The Time Off Request form contains a `manager`, `start_date`, and `end_date` field, all of which are required inputs to submit the form.
```

> After the form is submitted, another workflow step can be added to handle the information that was submitted in this form. For the Time Off Request workflow, after the requester submits the request, it will send a message to the manager to notify them of the request and allow them to approve or deny the request through the message. Since this functionality does not exist as a built-in function, it requires a custom function in order to execute the desired functionality. In our workflow, we'll call it `ApprovalFunction`:

Reviewer: ```suggestion
After the form is submitted, another workflow step handles the information that was submitted in this form. In our example, after the requester submits, we want to send a message to their manager to notify them of the request and allow them to approve or deny the request through the message. Since this functionality does not exist as a built-in function, let's write our own custom function to do this work. In our workflow, we'll call it `ApprovalFunction`:
```

      },
    },
    required: [],
  },
});

const step1 = TimeOffWorkflow.addStep(
  Schema.slack.functions.OpenForm,
  {
    title: 'Request Time Off',
    submit_label: 'Request',
    description: 'Please describe your request',
    interactivity: TimeOffWorkflow.inputs.interactivity,
    fields: {
      required: ['manager', 'start_date', 'end_date'],
      elements: [
        {
          name: 'manager',
          title: 'Manager',
          type: Schema.slack.types.user_id,
        },
        {
          name: 'start_date',
          title: 'Start Date',
          type: Schema.slack.types.timestamp,
        },
        {
          name: 'end_date',
          title: 'End Date',
          type: Schema.slack.types.timestamp,
        },
      ],
    },
  },
);

TimeOffWorkflow.addStep(ApprovalFunction, {
  employee: TimeOffWorkflow.inputs.interactivity.interactor.id,
  manager: step1.outputs.fields.manager,
  start_date: step1.outputs.fields.start_date,
  end_date: step1.outputs.fields.end_date,
});

module.exports = { TimeOffWorkflow };
```

To explore more about Workflows and dive into creating new ones, visit the guide [here](/bolt-js/future/workflows).

---

> ### Triggers {#triggers}

Reviewer: ```suggestion
### Link Triggers {#triggers}
```
This is a link trigger, and maybe we send users to the more general page on triggers?

Reviewer 2: sounds good! I'll keep in the blurb about link triggers at the top and then cut out everything up to "To learn more about Triggers and create new ones, visit the guide [here]"

        type: Schema.slack.types.user_id,
        description: 'Requester',
      },
      manager: {
        type: Schema.slack.types.user_id,
        description: 'Manager',
      },
      start_date: {
        type: Schema.slack.types.timestamp,
        description: 'Start Date',
      },
      end_date: {
        type: Schema.slack.types.timestamp,
        description: 'End Date',
      },
    },
    required: [
      'employee',
      'manager',
      'start_date',
      'end_date',
    ],
  },
  output_parameters: {
    properties: {},
    required: [],
  },
});

module.exports = { ApprovalFunction };
```
Within `ApprovalFunction`'s definition, we are declaring expected input and output parameters. We are also adding some additional metadata about our function, like a `title` and `description`.

Parameters for the function are typed using [types from Slack's schema](https://api.slack.com/future/types), such as having the `user_id` type for the `employee` parameter.

Once the function is defined, we need to create a listener for it to perform any desired functionality. In this case, our desired functionality is to send the inputted manager the time off request as a message.

> The function listener is declared in `listeners/functions/request-approval.js`:

Reviewer: The code sample below is on the longer side, should we remove it and link to the file instead for easier maintainability?

Reviewer 2: sounds good to me!
